File: leetcode/linkedlist/24.Swap-Nodes-in-Pairs/24.Swap-Nodes-in-Pairs.go
package leetcode

import (
	"github.com/yuancf1024/algorithm-go/structures"
)

type ListNode = structures.ListNode

/* 24. Swap Nodes in Pairs
给你一个链表，两两交换其中相邻的节点，并返回交换后链表的头节点。
你必须在不修改节点内部的值的情况下完成本题（即，只能进行节点交换）。

两两相邻的元素，翻转链表
*/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
//

// 方法1：直接按照题意做 不太好理解
func swapPairs(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }
	dummy := &ListNode{Next: head} // 哨兵节点
	for pt := dummy; pt != nil && pt.Next != nil && pt.Next.Next != nil; {
		pt, pt.Next, pt.Next.Next, pt.Next.Next.Next = pt.Next, pt.Next.Next, pt.Next.Next.Next, pt.Next
	}
	return dummy.Next
}

/* 按照题意做即可。
*/

// 方法2：递归实现
func swapPairs_v2(head *ListNode) *ListNode {
	// 思路：将链表翻转转化为一个子问题，然后通过递归的方式依次解决
	// 先翻转两个，然后将后面的节点继续这样反转，然后将这些翻转后的节点连接起来
	return helper(head)
}

func helper(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	// 保存下一阶段的头指针
	nextHead := head.Next.Next
	// 翻转当前阶段指针
	next := head.Next
	next.Next = head
	head.Next = helper(nextHead)
	return next
}

File: leetcode-cpp/24.Swap-Nodes-in-Pairs/24.Swap-Nodes-in-Pairs.cpp
/* 24. Swap Nodes in Pairs
给你一个链表，两两交换其中相邻的节点，并返回交换后链表的头节点。
你必须在不修改节点内部的值的情况下完成本题（即，只能进行节点交换）。

示例 1：
输入：head = [1,2,3,4]
输出：[2,1,4,3]

示例 2：
输入：head = []
输出：[]

示例 3：
输入：head = [1]
输出：[1]
 
提示：

链表中节点的数目在范围 [0, 100] 内
0 <= Node.val <= 100

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 参考代码随想录
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummyHead = new ListNode(0); // 设置一个虚拟头节点
        dummyHead->next = head; // 将虚拟头节点指向head，这样方便后面做删除操作
        ListNode* cur = dummyHead;
        while (cur->next != nullptr && cur->next->next != nullptr) {
            ListNode* tmp = cur->next; // 记录临时节点
            ListNode* tmp1 = cur->next->next->next; // 记录临时节点

            cur->next = cur->next->next; // 步骤1
            cur->next->next = tmp; // 步骤2
            cur->next->next->next = tmp1; // 步骤3

            cur = cur->next->next; // 移动两位，准备下一轮交换
        }
        return dummyHead->next;
    }
};

/*
思路：

这道题目正常模拟就可以了。

建议使用虚拟头结点，这样会方便很多，要不然每次针对头结点
（没有前一个指针指向头结点），还要单独处理。

接下来就是交换相邻两个元素了，此时一定要画图，不画图，操作多个指针很容易乱，
而且要操作的先后顺序

初始时，cur指向虚拟头结点，然后进行如下三步：

cur->1->2->3->4->

1. cur->2
2. 2->1
3. 1->3

操作之后，链表如下：
cur->2->1->3->4->

*/